filename validation ConstraintViolationException

With dv-webloader, one may drop a folder containing thousands of files and begin upload. dv-webloader will upload them. The user sees:
![Screenshot 2023-08-11 at 10 46 10](https://github.com/gdcc/dvwebloader/assets/12831666/5501dc3a-3aa3-4300-9a53-b4d9e8488e27)

Then, in the javascript console, things go splat:
![Screenshot 2023-08-11 at 10 44 41](https://github.com/gdcc/dvwebloader/assets/12831666/39cddb68-1e89-4dae-afd5-1f641052d2d6)

Payara's server.log sez
```
Constraint violation found in FileMetadata. File Name cannot contain any of the following characters:  / : * ? " < > | ; # . The invalid value is "filename".

  javax.ejb.EJBException: One or more Bean Validation constraints were violated while executing Automatic Bean Validation on callback event: prePersist for class: edu.harvard.iq.dataverse.FileMetadata. Please refer to the embedded constr
aint violations for details.
        at com.sun.ejb.containers.EJBContainerTransactionManager.processSystemException(EJBContainerTransactionManager.java:723)
        at com.sun.ejb.containers.EJBContainerTransactionManager.completeNewTx(EJBContainerTransactionManager.java:652)
        at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:482)
        at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4601)
        at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2134)
        at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2104)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:220)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:90)
        at com.sun.proxy.$Proxy377.submit(Unknown Source)
        at edu.harvard.iq.dataverse.__EJB31_Generated__EjbDataverseEngineInner__Intf____Bean__.submit(Unknown Source)
        at edu.harvard.iq.dataverse.EjbDataverseEngine.submit(EjbDataverseEngine.java:272)
        at jdk.internal.reflect.GeneratedMethodAccessor1337.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:588)
        at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:408)
        at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4835)
        at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:665)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:834)
        at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:615)
        at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
        at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
        at jdk.internal.reflect.GeneratedMethodAccessor150.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:888)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:833)
        at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:615)
        at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:72)
        at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
        at jdk.internal.reflect.GeneratedMethodAccessor144.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:888)
        at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:833)
        at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:375)
        at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4807)
        at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4795)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
        at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:90)
        at com.sun.proxy.$Proxy405.submit(Unknown Source)
        at edu.harvard.iq.dataverse.__EJB31_Generated__EjbDataverseEngine__Intf____Bean__.submit(Unknown Source)
        at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.addFiles(AddReplaceFileHelper.java:2148)
        at edu.harvard.iq.dataverse.api.Datasets.addFilesToDataset(Datasets.java:3467)
Caused by: javax.validation.ConstraintViolationException: One or more Bean Validation constraints were violated while executing Automatic Bean Validation on callback event: prePersist for class: edu.harvard.iq.dataverse.FileMetadata. Please refer to the embedded constraint violations for details.
        at org.eclipse.persistence.internal.jpa.metadata.listeners.BeanValidationListener.validateOnCallbackEvent(BeanValidationListener.java:124)
        at org.eclipse.persistence.internal.jpa.metadata.listeners.BeanValidationListener.prePersist(BeanValidationListener.java:86)
        at org.eclipse.persistence.descriptors.DescriptorEventManager.notifyListener(DescriptorEventManager.java:746)
        at org.eclipse.persistence.descriptors.DescriptorEventManager.notifyEJB30Listeners(DescriptorEventManager.java:689)
        at org.eclipse.persistence.descriptors.DescriptorEventManager.executeEvent(DescriptorEventManager.java:233)
        at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.registerNewObjectClone(UnitOfWorkImpl.java:4468)
        at org.eclipse.persistence.internal.sessions.RepeatableWriteUnitOfWork.cloneAndRegisterNewObject(RepeatableWriteUnitOfWork.java:613)
        at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalRegisterObject(UnitOfWorkImpl.java:3045)
        at org.eclipse.persistence.internal.sessions.MergeManager.registerObjectForMergeCloneIntoWorkingCopy(MergeManager.java:1102)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChangesOfCloneIntoWorkingCopy(MergeManager.java:575)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChanges(MergeManager.java:324)
        at org.eclipse.persistence.mappings.CollectionMapping.mergeIntoObject(CollectionMapping.java:1650)
        at org.eclipse.persistence.internal.descriptors.ObjectBuilder.mergeIntoObject(ObjectBuilder.java:4202)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChangesOfCloneIntoWorkingCopy(MergeManager.java:612)
        at org.eclipse.persistence.internal.sessions.MergeManager.mergeChanges(MergeManager.java:324)
        at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.mergeCloneWithReferences(UnitOfWorkImpl.java:3637)
        at org.eclipse.persistence.internal.sessions.RepeatableWriteUnitOfWork.mergeCloneWithReferences(RepeatableWriteUnitOfWork.java:389)
        at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.mergeCloneWithReferences(UnitOfWorkImpl.java:3597)
        at org.eclipse.persistence.internal.jpa.EntityManagerImpl.mergeInternal(EntityManagerImpl.java:648)
        at org.eclipse.persistence.internal.jpa.EntityManagerImpl.merge(EntityManagerImpl.java:625)
        at com.sun.enterprise.container.common.impl.EntityManagerWrapper.merge(EntityManagerWrapper.java:307)
        at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetVersionCommand.execute(UpdateDatasetVersionCommand.java:129)
        at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetVersionCommand.execute(UpdateDatasetVersionCommand.java:30)
        at edu.harvard.iq.dataverse.EjbDataverseEngineInner.submit(EjbDataverseEngineInner.java:36)
 ```

It would be great if dv-webloader could either percolate the user-friendly error back to the dv-webloader UI, or perhaps pre-emptively search for such characters before attempting to upload files?

--- a/src/css/dvwebloader.css
+++ b/src/css/dvwebloader.css
@@ -98,3 +98,9 @@ h1 {
    background-color: bisque;
     padding: 10px;
 }
+#content-warnings div.warn {
+   margin: 10px;
+   border-color: chocolate;
+   border-width: 2px;
+   border-style: solid;
+}

--- a/src/dvwebloader.html
+++ b/src/dvwebloader.html
@@ -27,7 +27,7 @@
             </div>
             <div id="filelist"></div>
             <div id="credit">
-                <a href='https://github.com/gdcc/dvwebloader' target='_blank'>DVWebloader v0.3</a><span id="sponsor-text">, development sponsored by UiT/DataverseNO</span>
+                <a href='https://github.com/gdcc/dvwebloader' target='_blank'>DVWebloader v0.4alpha1</a><span id="sponsor-text">, development sponsored by UiT/DataverseNO</span>
             </div>
             <script>
                 var input = document

--- a/src/dvwebloader.html
+++ b/src/dvwebloader.html
@@ -23,6 +23,7 @@
             <div id="top">
                 <label for='files' class='button'><span id="select-dir-text">Select a Directory</span><input type="file" id="files" name="files[]" multiple webkitdirectory style="display: none" /></label>
                 <div id='messages'></div>
+                <div id='content-warnings'></div>
             </div>
             <div id="filelist"></div>
             <div id="credit">

--- a/src/js/fileupload2.js
+++ b/src/js/fileupload2.js
@@ -327,8 +327,8 @@ async function retrieveDatasetInfo() {
                 let df = entry.dataFile;
                 let convertedFile = false;
                 if (("originalFileFormat" in df)
-                    && (!df.contentType === df.originalFileFormat)) {
-                    console.log("The file named " + df.getString("filename")
+                    && (!(df.contentType === df.originalFileFormat))) {
+                    console.log("The file named " + df.filename
                         + " on the server was created by Dataverse's ingest process from an original uploaded file");
                     convertedFile = true;
                 }
@@ -745,7 +745,7 @@ var fileUpload = class fileUploadClass {
 }
     ;
 function removeExtension(name) {
-    let extIndex = name.indexOf(".");
+    let extIndex = name.lastIndexOf(".");
     let sepIndex = name.indexOf('/');
     if (extIndex > sepIndex) {
         return name.substring(0, extIndex);

--- a/src/js/fileupload2.js
+++ b/src/js/fileupload2.js
@@ -238,6 +238,17 @@ function addMessage(type, key, ...keyArgs) {
     $('#messages').html('')
         .append($('<div/>').addClass(type).html(msg));
 }
+function addContentWarning(key, ...keyArgs) {
+    let msg = getLocalizedString(dvLocale, key);
+    
+    if(keyArgs && Array.isArray(keyArgs)) {
+        for (var i = 0; i < keyArgs.length; i++) {
+            msg = msg.replaceAll('{'+i+'}',keyArgs[i]);
+        }
+    }
+    $('#content-warnings').html('')
+        .append($('<div/>').addClass('warn').html(msg));
+}
 
 async function populatePageMetadata(data) {
     var mdFields = data.metadataBlocks.citation.fields;
@@ -765,8 +776,8 @@ function queueFileForDirectUpload(file) {
     let path =origPath.substring(0, origPath.length - file.name.length);
     let badPath = (path.match(/^[\w\-\.\\\/ ]*$/)===null);
     if(badPath) {
-      if($('.warn').length==0) {
-        addMessage('warn', 'msgRequiredPathOrFileNameChange');
+      if($('#content-warnings.warn').length==0) {
+        addContentWarning('msgRequiredPathOrFileNameChange');
       }
       //Munge path according to rules
       path = path.replace(/[^\w\-\.\\\/ ]+/g,'_');
@@ -799,8 +810,8 @@ function queueFileForDirectUpload(file) {
     }
     let badChars = !(fUpload.file.name.match(/[:<>;#\/"*|?\\]/)===null);
     if(badChars) {
-      if($('.warn').length==0) {
-          addMessage('warn', 'msgRequiredPathOrFileNameChange');
+      if($('#content-warnings.warn').length==0) {
+          addContentWarning('msgRequiredPathOrFileNameChange');
       }
     }
     row.append($('<input/>').prop('type', 'checkbox').prop('id', 'file_' + fUpload.id).prop('checked', send));

--- a/src/js/fileupload2.js
+++ b/src/js/fileupload2.js
@@ -266,52 +266,98 @@ async function populatePageMetadata(data) {
 }
 
 async function retrieveDatasetInfo() {
-    $.ajax({
-        url: siteUrl + '/api/datasets/:persistentId/versions/:latest?persistentId=' + datasetPid,
-        headers: { "X-Dataverse-key": apiKey },
-        type: 'GET',
-        context: this,
-        cache: false,
-        dataType: "json",
-        processData: false,
-        success: function(body, statusText, jqXHR) {
-            console.log(body);
-            let data = body.data;
-            console.log(data);
-            populatePageMetadata(data);
-            if (data.files !== null) {
-                existingFiles = {};
-                convertedFileNameMap = {};
-                for (let i = 0; i < data.files.length; i++) {
-                    let entry = data.files[i];
-                    let df = entry.dataFile;
-                    let convertedFile = false;
-                    if (("originalFileFormat" in df)
-                        && (!df.contentType === df.originalFileFormat)) {
-                        console.log("The file named " + df.getString("filename")
-                            + " on the server was created by Dataverse's ingest process from an original uploaded file");
-                        convertedFile = true;
-                    }
-                    let filepath = df.filename;
-                    if ('directoryLabel' in entry) {
-                        filepath = entry.directoryLabel + '/' + filepath;
-                    }
-                    console.log("Storing: " + filepath);
-                    existingFiles[filepath] = df.checksum;
-                    if (convertedFile) {
-                        convertedFileNameMap[removeExtension(filepath)] = filepath;
-                    }
+    try {
+        // First, check for dataset locks
+        const locksResponse = await $.ajax({
+            url: siteUrl + '/api/datasets/:persistentId/locks?persistentId=' + datasetPid,
+            headers: { "X-Dataverse-key": apiKey },
+            type: 'GET',
+            cache: false,
+            dataType: "json"
+        });
+
+        let isLockedInReview = false;
+        if (locksResponse.data && locksResponse.data.length > 0) {
+            isLockedInReview = locksResponse.data.some(lock => lock.lockType === "InReview");
+        }
+
+        // If locked but not InReview, disable upload
+        if (locksResponse.data && locksResponse.data.length > 0 && !isLockedInReview) {
+            addMessage('error', 'msgDatasetLocked');
+            disableUploadFunctionality();
+            return;
+        }
+
+        // If locked InReview, check user permissions
+        if (isLockedInReview) {
+            const permissionsResponse = await $.ajax({
+                url: siteUrl + '/api/datasets/:persistentId/userPermissions?persistentId=' + datasetPid,
+                headers: { "X-Dataverse-key": apiKey },
+                type: 'GET',
+                cache: false,
+                dataType: "json"
+            });
+
+            if (!permissionsResponse.data || !permissionsResponse.data.canPublishDataset) {
+                addMessage('error', 'msgDatasetLockedInReview');
+                disableUploadFunctionality();
+                return;
+            }
+        }
+
+        // If not locked or user has permission, proceed with retrieving dataset info
+        const datasetResponse = await $.ajax({
+            url: siteUrl + '/api/datasets/:persistentId/versions/:latest?persistentId=' + datasetPid,
+            headers: { "X-Dataverse-key": apiKey },
+            type: 'GET',
+            cache: false,
+            dataType: "json"
+        });
+
+        console.log(datasetResponse);
+        let data = datasetResponse.data;
+        console.log(data);
+
+        populatePageMetadata(data);
+        if (data.files !== null) {
+            existingFiles = {};
+            convertedFileNameMap = {};
+            for (let i = 0; i < data.files.length; i++) {
+                let entry = data.files[i];
+                let df = entry.dataFile;
+                let convertedFile = false;
+                if (("originalFileFormat" in df)
+                    && (!df.contentType === df.originalFileFormat)) {
+                    console.log("The file named " + df.getString("filename")
+                        + " on the server was created by Dataverse's ingest process from an original uploaded file");
+                    convertedFile = true;
+                }
+                let filepath = df.filename;
+                if ('directoryLabel' in entry) {
+                    filepath = entry.directoryLabel + '/' + filepath;
+                }
+                console.log("Storing: " + filepath);
+                existingFiles[filepath] = df.checksum;
+                if (convertedFile) {
+                    convertedFileNameMap[removeExtension(filepath)] = filepath;
                 }
             }
-            $('#files').prop('disabled', false);
-            addMessage('info', 'msgReadyToStart');
-        },
-        error: function(jqXHR, textStatus, errorThrown) {
-            console.log('Failure: ' + jqXHR.status);
-            console.log('Failure: ' + errorThrown);
         }
-    });
+        $('#files').prop('disabled', false);
+        addMessage('info', 'msgReadyToStart');
+    } catch (error) {
+        console.log('Error:', error);
+        addMessage('error', 'msgErrorRetrievingDataset');
+    }
 }
+
+function disableUploadFunctionality() {
+    $('#files').prop('disabled', true);
+    $('.file-selection-buttons').hide();
+    $('label.button').hide();
+    // Disable any other relevant UI elements
+}
+
 //Not used in dvwebloader
 function setupDirectUpload(enabled) {
     if (enabled) {
@@ -867,7 +913,6 @@ async function uploadFileDirectly(urls, storageId, filesize) {
 
 
 
-
 function removeErrors() {
     var errors = document.getElementsByClassName("ui-fileupload-error");
     for (let i = errors.length - 1; i >= 0; i--) {
@@ -1136,7 +1181,7 @@ function getChecksum(blob, cbProgress) {
                 checksumAlg = CryptoJS.algo.SHA512.create();
                 break;
             default:
-                console.log('$(checksumAlgName) is not supported, using MD5 as the checksumAlg checksum Algorithm');
+                console.log('$(checksumAlgName) is not supported, using MD5 as the checksum Algorithm');
                 checksumAlg = CryptoJS.algo.MD5.create();
         }
         readChunked(blob, (chunk, offs, total) => {

--- a/src/js/fileupload2.js
+++ b/src/js/fileupload2.js
@@ -227,25 +227,26 @@ function initTranslation() {
 function initSpanTxt(htmlId, key) {
     $('#'+htmlId).text(getLocalizedString(dvLocale, key));
 }
-function addMessage(type, key, ...keyArgs) {
+
+function formatMessage(key, keyArgs) {
     let msg = getLocalizedString(dvLocale, key);
     
     if(keyArgs && Array.isArray(keyArgs)) {
         for (var i = 0; i < keyArgs.length; i++) {
-            msg = msg.replaceAll('{'+i+'}',keyArgs[i]);
+            msg = msg.replaceAll('{'+i+'}', keyArgs[i]);
         }
     }
+    return msg;
+}
+
+function addMessage(type, key, ...keyArgs) {
+    let msg = formatMessage(key, keyArgs);
     $('#messages').html('')
         .append($('<div/>').addClass(type).html(msg));
 }
+
 function addContentWarning(key, ...keyArgs) {
-    let msg = getLocalizedString(dvLocale, key);
-    
-    if(keyArgs && Array.isArray(keyArgs)) {
-        for (var i = 0; i < keyArgs.length; i++) {
-            msg = msg.replaceAll('{'+i+'}',keyArgs[i]);
-        }
-    }
+    let msg = formatMessage(key, keyArgs);
     $('#content-warnings').html('')
         .append($('<div/>').addClass('warn').html(msg));
 }

--- a/src/js/lang.js
+++ b/src/js/lang.js
@@ -23,6 +23,9 @@ const translations = {
         msgDeselectAll: "Deselect all",
         msgMaxFilesExceeded: "Maximum number of files exceeded. Uncheck some files to enable upload.",
         msgMaxFilesReached: "Maximum number of files reached. Uncheck some files before selecting more.",
+        msgDatasetLocked: "This dataset is currently locked. File uploads are not allowed at this time.",
+        msgDatasetLockedInReview: "This dataset is currently In Review. You are not allowed to upload files at this time.",
+        msgErrorRetrievingDataset: "An error occurred while retrieving dataset information."
     },
     fr: {
         title: "Envoi d'un dossier",
@@ -47,6 +50,9 @@ const translations = {
         msgDeselectAll: "Tout désélectionner",
         msgMaxFilesExceeded: "Nombre maximal de fichiers dépassé. Décochez certains fichiers pour activer le transfert.",
         msgMaxFilesReached: "Nombre maximal de fichiers atteint. Décochez certains fichiers avant d'en sélectionner d'autres.",
+        msgDatasetLocked: "Ce jeu de données est actuellement verrouillé. Les envois de fichiers ne sont pas autorisés pour le moment.",
+        msgDatasetLockedInReview: "Ce jeu de données est actuellement en cours de révision. Vous n'êtes pas autorisé à télécharger des fichiers pour le moment.",
+        msgErrorRetrievingDataset: "Une erreur s'est produite lors de la récupération des informations du jeu de données."
     },
 };
 

